File: dao/src/main/scala/farjs/domain/dao/HistoryKindDao.scala
package farjs.domain.dao

import farjs.domain._
import scommons.websql.io.dao.CommonDao

import scala.concurrent.Future

class HistoryKindDao(val ctx: FarjsDBContext) extends CommonDao {

  import ctx._

  private val tableName = "history_kinds"

  def getAll: Future[Seq[HistoryKindEntity]] = {
    ctx.performIO(ctx.runQuery(
      sql = s"SELECT id, name FROM $tableName ORDER BY id",
      extractor = HistoryKindEntity.tupled
    ))
  }

  def upsert(entity: HistoryKindEntity): Future[HistoryKindEntity] = {
    val q = for {
      _ <- ctx.runAction(
        sql = s"INSERT INTO $tableName (name) VALUES (?) ON CONFLICT (name) DO NOTHING",
        args = entity.name
      )
      res <- ctx.runQuery(
        sql = s"SELECT id, name FROM $tableName WHERE name = ?",
        args = entity.name,
        extractor = HistoryKindEntity.tupled
      )
    } yield res.head

    ctx.performIO(q)
  }
  
  def deleteAll(): Future[Long] = {
    ctx.performIO(ctx.runAction(s"DELETE FROM $tableName"))
  }
}

File: dao/HistoryKindDao.mjs
import Database from "@farjs/better-sqlite3-wrapper";

/**
 * @typedef {{
 *  readonly id: number;
 *  readonly name: string;
 * }} HistoryKindEntity
 */

/**
 * @typedef {{
 *  getAll(): Promise<readonly HistoryKindEntity[]>;
 *  upsert(entity: HistoryKindEntity): Promise<HistoryKindEntity>;
 *  deleteAll(): Promise<void>;
 * }} HistoryKindDao
 */

const tableName = "history_kinds";

/**
 *
 * @param {Database.Database} db
 * @returns {HistoryKindDao}
 */
function HistoryKindDao(db) {
  return {
    getAll: async () =>
      db.transaction(() => {
        const query = db.prepare(
          /* sql */ `SELECT id, name FROM ${tableName} ORDER BY id;`,
        );
        return /** @type {HistoryKindEntity[]} */ (query.all());
      })(),

    upsert: async (entity) =>
      db.transaction(() => {
        db.prepare(
          /* sql */ `INSERT INTO ${tableName} (name) VALUES (?) ON CONFLICT (name) DO NOTHING;`,
        ).run(entity.name);

        const query = db.prepare(
          /* sql */ `SELECT id, name FROM ${tableName} WHERE name = ?;`,
        );
        return /** @type {HistoryKindEntity} */ (query.get(entity.name));
      })(),

    deleteAll: async () =>
      db.transaction(() => {
        db.prepare(/* sql */ `DELETE FROM ${tableName};`).run();
      })(),
  };
}

export default HistoryKindDao;
